core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareAsyncProcessor.java

```diff
@@ -45,12 +51,22 @@ public boolean canProcess(Request request, GenericType<?> resultType) {
   }
 
   @Override
-  public RequestHandler<PrepareRequest, CompletionStage<PreparedStatement>> newHandler(
+  public CompletionStage<PreparedStatement> process(
       PrepareRequest request,
       DefaultSession session,
       InternalDriverContext context,
       String sessionLogPrefix) {
-    return new CqlPrepareAsyncHandler(
-        request, preparedStatementsCache, session, context, sessionLogPrefix);
+
+    try {
+      return cache.get(
+          request,
+          () -> new CqlPrepareHandler(request, session, context, sessionLogPrefix).handle());
```

Comment:
```suggestion
          () -> {
            CompletableFuture<PreparedStatement> future =
                new CqlPrepareHandler(request, session, context, sessionLogPrefix).handle();

            // invalidate this entry if it failed so subsequent requests may try preparing again.
            future.whenComplete(
                (statement, ex) -> {
                  if (ex != null) {
                    cache.invalidate(request);
                  }
                });
            return future;
          });
```

Would something like this work? Cache entry is invalidated if future fails.  It makes my test pass but i'm not sure if the right fix.  

One possible concern is if somehow the future completes before the item is cached, which I suppose is possible, yet unlikely since the future completes on a different thread.

Reply:
Good catch, but I share your concern about the race condition in your suggested fix: the window is small, but if it ever happens the error is cached forever (or at least until eviction), which is not acceptable. 

We need to be sure that the entry exists before we check the future for failure. How about doing it after every lookup:
```java
      CompletableFuture<PreparedStatement> result =
          cache.get(
              request,
              () -> new CqlPrepareHandler(request, session, context, sessionLogPrefix).handle());
      if (result.isCompletedExceptionally()) {
        cache.invalidate(request);
      }
      return result;
```

Reply:
If the Future will complete with the exception after the if check, our `isCompletedExceptionally()` will not work properly (and this is very probable):
```
    CompletableFuture<Boolean> result =
        new CompletableFuture<>();

    if (result.isCompletedExceptionally()) {
      System.out.println("Action on ex");
    }

    result.completeExceptionally(new Exception("a"));
```
we would need that order:
```
CompletableFuture<Boolean> result =
        new CompletableFuture<>();
    result.completeExceptionally(new Exception("a"));

    if (result.isCompletedExceptionally()) {
      System.out.println("Action on ex");
    }
```
Maybe we can chain it using `completeExceptionally()` something like:
```
  CompletableFuture<Boolean> result =
        new CompletableFuture<>();

    result.exceptionally(throwable -> {
      System.out.println("on error"); //cache invalidate here
      return true;
    });

    result.completeExceptionally(new Exception("a"));
```
but seems that still, that problem occurs: `if somehow the future completes before the item is cached, `
Maybe in a guava cache loader, we can add a condition that should not cache if CompletableFuture returns false (or exception)?

Reply:
Yes, for the initial call the future will probably not be complete yet when the `if` block is reached, it will be detected on the next call to `prepare()`. That sounded reasonable to me but on second thought, if you're doing serial calls from the same thread, it means you have to retry 2 times which is not great.
Maybe we could keep both checks: @tolbertam's `whenComplete` in the cache loader, and this one as a last resort if the future completed before the cache entry was created (which frankly will probably be very rare).

> Or we should not cache CompletableFuture at all - but the actual result?

This would make `prepareAsync()` blocking.

Reply:
Even better: put a standalone future in the cache, and only then call the handler and propagate the result.
```java
CompletableFuture<PreparedStatement> result = cache.getIfPresent(request);
if (result == null) {
  CompletableFuture<PreparedStatement> mine = new CompletableFuture<>();
  result = cache.get(request, () -> mine);
  if (result == mine) {
    new CqlPrepareHandler(request, session, context, sessionLogPrefix)
        .handle()
        .whenComplete(
            (preparedStatement, error) -> {
              if (error != null) {
                mine.completeExceptionally(error);
                cache.invalidate(request); // Make sure failure isn't cached indefinitely
              } else {
                mine.complete(preparedStatement);
              }
            });
  }
}
return result;
```

Reply:
Implemented @olim7t suggestion and squashed the commits, keeping 3 separate ones as requested.

Reply:
because the path `(result != mine)` is not covered it is prone to the same problem as the previous solution, isn't it?

Reply:
Also, there is possiblity that two/or more threads will trigger compute because of the check `result == null`. Are we ok with that?

Reply:
Yes, because only one thread will succeed in putting `mine` in the cache.

Reply:
If `result != mine`, then another thread has beat us to putting its own future in the cache, and triggered the computation. We don't trigger the computation on our side, and return its future.

Two or more threads can enter `result == null`, but only one can "win" the cache initialization, i.e. enter the `result == mine` block. There is a race, but the only consequence is that the losing threads have created an extra `mine` future that will be immediately discarded.